cql3/expr/to_restriction.cc

```diff
@@ -20,14 +20,164 @@
+    r->expression = binary_operator(std::move(lhs_col), oper, std::move(prepared_rhs), order);
+    return r;
+}
+
+::shared_ptr<restrictions::restriction> new_subscripted_column_restriction(subscript&& lhs_sub,
+                                                                           oper_t oper,
+                                                                           expression&& prepared_rhs,
+                                                                           comparison_order order,
+                                                                           const schema& schema) {
+    const column_value& sub_col = get_subscripted_column(lhs_sub);
+    validate_single_column_relation(sub_col, oper, schema, true);
+
+    ::shared_ptr<restrictions::restriction> r = ::make_shared<restrictions::single_column_restriction>(*sub_col.col);
+    r->expression = binary_operator(std::move(lhs_sub), oper, std::move(prepared_rhs));
+    return r;
+}
+
+} // anonymous namespace
+
 ::shared_ptr<restrictions::restriction> to_restriction(const expression& e,
                                                        data_dictionary::database db,
                                                        schema_ptr schema,
                                                        prepare_context& ctx) {
-    throw std::runtime_error(fmt::format("Unimplemented! {}:{}", __FILE__, __LINE__));
+    if (is<constant>(e)) {
+        throw exceptions::invalid_request_exception("Constant value by itself is not supported as a restriction yet");
+    }
+
+    if (!is<binary_operator>(e)) {
+        throw exceptions::invalid_request_exception("Restriction must be a binary operator");
+    }
+    binary_operator binop_to_prepare = as<binary_operator>(e);
+
+    // Convert single element IN relation to an EQ relation
+    if (binop_to_prepare.op == oper_t::IN && is<collection_constructor>(binop_to_prepare.rhs)) {
+        const std::vector<expression>& elements = as<collection_constructor>(binop_to_prepare.rhs).elements;
+        if (elements.size() == 1) {
+            binop_to_prepare.op = oper_t::EQ;
+            binop_to_prepare.rhs = elements[0];
+        }
+    }
+
+    binary_operator prepared_binop = prepare_binary_operator(binop_to_prepare, db, schema, ctx);
+
+    const column_value* lhs_pk_col_search_res = find_in_expression<column_value>(prepared_binop.lhs,
+        [](const column_value& col) {
+            return col.col->is_partition_key();
+        }
+    );
+    auto reset_processing_pk_column = defer([&ctx] () noexcept { ctx.set_processing_pk_restrictions(false); });
+    if (lhs_pk_col_search_res != nullptr) {
+        ctx.set_processing_pk_restrictions(true);
+    }
+    fill_prepare_context(prepared_binop.lhs, ctx);
+    fill_prepare_context(prepared_binop.rhs, ctx);
+
+    if (prepared_binop.op == oper_t::NEQ) {
+        throw exceptions::invalid_request_exception(format("Unsupported \"!=\" relation: {}", prepared_binop));
+    }
```

Comment:
Wait, we don't support this?

Reply:
We don't
```cql
cqlsh> CREATE KEYSPACE IF NOT EXISTS ks WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 1};
cqlsh> USE ks;
cqlsh:ks> CREATE TABLE tab (p int primary key);
cqlsh:ks> SELECT * FROM tab WHERE p != 0 ALLOW FILTERING;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unsupported "!=" relation: p != 0"
```
Neither does cassandra

Here's the code responsible for it:
https://github.com/scylladb/scylla/blob/528ab5a5026e83c35a6a3b6d476230b692099a2a/cql3/relation.hh#L134

Reply:
So it's impossible to use it anywhere, and it's just in the grammar?

I don't mind, but it's odd.

Reply:
> So it's impossible to use it anywhere, and it's just in the grammar?

It looks like there is one place in `column_condtion` that has an if for something like `a != NULL`.
https://github.com/scylladb/scylla/blob/528ab5a5026e83c35a6a3b6d476230b692099a2a/cql3/column_condition.cc#L190
But that's the only actually possible use of this operator that I found.
Btw shouldn't this be `a IS NOT NULL`?

`is_satisfied_by` already has implementation for `!=`, so it would be probably easy to add support for it, at least with allow filtering But maybe it's better to wait until the end of expression rewrite and then something more efficient could be implemented.

Reply:
> > So it's impossible to use it anywhere, and it's just in the grammar?
> 
> It looks like there is one place in `column_condtion` that has an if for something like `a != NULL`.
> 
> https://github.com/scylladb/scylla/blob/528ab5a5026e83c35a6a3b6d476230b692099a2a/cql3/column_condition.cc#L190
> 
> 
> But that's the only actually possible use of this operator that I found.
> Btw shouldn't this be `a IS NOT NULL`?

Probably. Please add an issue.

> `is_satisfied_by` already has implementation for `!=`, so it would be probably easy to add support for it, at least with allow filtering But maybe it's better to wait until the end of expression rewrite and then something more efficient could be implemented.

Sure, it's low priority in any case.